extensions/gitpod/src/authentication.ts

```diff
@@ -0,0 +1,272 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { v4 as uuid } from 'uuid';
+import { Keychain } from './common/keychain';
+import { GitpodServer } from './gitpodServer';
+import Log from './common/logger';
+import { arrayEquals } from './common/utils';
+import { Disposable } from './common/dispose';
+import TelemetryReporter from './telemetryReporter';
+
+interface SessionData {
+	id: string;
+	account?: {
+		label?: string;
+		displayName?: string;
+		id: string;
+	};
+	scopes: string[];
+	accessToken: string;
+}
+
+export default class GitpodAuthenticationProvider extends Disposable implements vscode.AuthenticationProvider {
+	private _sessionChangeEmitter = new vscode.EventEmitter<vscode.AuthenticationProviderAuthenticationSessionsChangeEvent>();
+	private _logger: Log;
+	private _telemetry: TelemetryReporter;
+	private _gitpodServer: GitpodServer;
+	private _keychain: Keychain;
+
+	private _sessionsPromise: Promise<vscode.AuthenticationSession[]>;
+
+	constructor(private readonly context: vscode.ExtensionContext, logger: Log, telemetry: TelemetryReporter) {
+		super();
+
+		this._logger = logger;
+		this._telemetry = telemetry;
+		this._keychain = new Keychain(this.context, `gitpod.auth`, this._logger);
+
+		const serviceUrl = vscode.workspace.getConfiguration('gitpod').get<string>('host')!;
+		this._gitpodServer = new GitpodServer(serviceUrl, this._logger);
+		this._register(vscode.workspace.onDidChangeConfiguration(async e => {
+			if (e.affectsConfiguration('gitpod.host')) {
+				const serviceUrl = vscode.workspace.getConfiguration('gitpod').get<string>('host')!;
+				this._gitpodServer.dispose();
+				this._gitpodServer = new GitpodServer(serviceUrl, this._logger);
+
+				// Remove all sessions when gitpod.host changes
+				const sessions = await this._sessionsPromise;
+				await this._keychain.deleteToken();
```

Comment:
I'm not sure that it works properly.

When a user updates `host` we should delete session from keychain for old host? But only unload it from memory and load other to avoid auth again when a user switch back. So I will imagine that `Keychain` object accept `host` for each operation.

Could you confirm that switching between hosts does not require unnecessary auth again with current implementation?

Reply:
It requires auth again, why would users keep changing host? either they use saas or self hosted, changing host is such a cumbersome operation because you need to restart vscode (close all windows) again for it to take effect
And then it would be misleading you could enable setting sync with one host then change the host and settings sync will keep working and you would think it using the new host

Reply:
Yes, switch between gitpod.io and self hosted installation. It is a rare operation and then indeed we need to restart to apply changes. But I don't think annoy with requesting token each time is a way.

I wonder why is it so hard to implement properly? We did the same for local app and JetBrains and there a user can switch nicely for testing and self hosted. Let'd do proper here too?

Reply:
will take another